File: python/src/webex_message_handler/logger.py
"""Logger interface and built-in implementations."""

from __future__ import annotations

import logging
from typing import Protocol


class Logger(Protocol):
    """Protocol for logger implementations."""

    def debug(self, message: str, *args: object) -> None: ...
    def info(self, message: str, *args: object) -> None: ...
    def warning(self, message: str, *args: object) -> None: ...
    def error(self, message: str, *args: object) -> None: ...


class _NoopLogger:
    """Silent logger — all methods are no-ops."""

    def debug(self, message: str, *args: object) -> None:
        pass

    def info(self, message: str, *args: object) -> None:
        pass

    def warning(self, message: str, *args: object) -> None:
        pass

    def error(self, message: str, *args: object) -> None:
        pass


noop_logger: Logger = _NoopLogger()

# console_logger is a standard logging.Logger that writes to stderr
_std_logger = logging.getLogger("webex_message_handler")
if not _std_logger.handlers:
    _handler = logging.StreamHandler()
    _handler.setFormatter(logging.Formatter("%(levelname)s %(name)s: %(message)s"))
    _std_logger.addHandler(_handler)
    _std_logger.setLevel(logging.DEBUG)
console_logger: Logger = _std_logger

File: go/logger.go
package webexmessagehandler

import "log/slog"

// Logger defines the logging interface used by the handler.
// Compatible with *slog.Logger via the SlogAdapter.
type Logger interface {
	Debug(msg string, args ...any)
	Info(msg string, args ...any)
	Warn(msg string, args ...any)
	Error(msg string, args ...any)
}

// noopLogger is a silent logger — all methods are no-ops.
type noopLogger struct{}

func (noopLogger) Debug(string, ...any) {}
func (noopLogger) Info(string, ...any)  {}
func (noopLogger) Warn(string, ...any)  {}
func (noopLogger) Error(string, ...any) {}

// NoopLogger returns a silent logger.
func NoopLogger() Logger { return noopLogger{} }

// SlogLogger wraps a *slog.Logger to implement the Logger interface.
type SlogLogger struct {
	L *slog.Logger
}

func (s *SlogLogger) Debug(msg string, args ...any) { s.L.Debug(msg, args...) }
func (s *SlogLogger) Info(msg string, args ...any)  { s.L.Info(msg, args...) }
func (s *SlogLogger) Warn(msg string, args ...any)  { s.L.Warn(msg, args...) }
func (s *SlogLogger) Error(msg string, args ...any) { s.L.Error(msg, args...) }

// NewSlogLogger creates a Logger from a *slog.Logger.
func NewSlogLogger(l *slog.Logger) Logger {
	return &SlogLogger{L: l}
}
